crates/pipeline_manager/src/api/services/probe.rs

```diff
@@ -0,0 +1,235 @@
+    Success,
+    /// Probe has finished running and it failed.
+    Failure,
+}
+
+impl TryFrom<String> for ServiceProbeStatus {
+    type Error = DBError;
+    fn try_from(value: String) -> Result<Self, DBError> {
+        match value.as_str() {
+            "pending" => Ok(Self::Pending),
+            "running" => Ok(Self::Running),
+            "success" => Ok(Self::Success),
+            "failure" => Ok(Self::Failure),
+            _ => Err(DBError::unknown_service_probe_status(value)),
+        }
+    }
+}
+
+impl From<ServiceProbeStatus> for &'static str {
+    fn from(val: ServiceProbeStatus) -> Self {
+        match val {
+            ServiceProbeStatus::Pending => "pending",
+            ServiceProbeStatus::Running => "running",
+            ServiceProbeStatus::Success => "success",
+            ServiceProbeStatus::Failure => "failure",
+        }
+    }
+}
+
+/// Enumeration of all possible service probe requests.
+#[derive(Debug, Serialize, Deserialize, ToSchema, Clone, Eq, PartialEq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+#[serde(rename_all = "snake_case")]
+pub enum ServiceProbeRequest {
+    /// General test connectivity request.
+    /// Every service must implement this at least.
+    TestConnectivity,
+    /// Probe request specific to a Kafka service.
+    Kafka(KafkaServiceProbeRequest),
+}
+
+impl ServiceProbeRequest {
+    /// Unique service request type used for classification.
+    pub fn request_type(&self) -> String {
+        match self {
+            ServiceProbeRequest::TestConnectivity => "test_connectivity".to_string(),
+            ServiceProbeRequest::Kafka(req) => match req {
+                KafkaServiceProbeRequest::GetTopics => "kafka_get_topics".to_string(),
+            },
+        }
+    }
+
+    /// Deserialize from provided YAML.
+    pub fn from_yaml_str(s: &str) -> Self {
+        serde_yaml::from_str(s).unwrap()
+    }
+
+    /// Serialize to YAML for storage.
+    pub fn to_yaml(&self) -> String {
+        serde_yaml::to_string(&self).unwrap()
+    }
+}
+
+/// Enumeration of all possible service probe success responses.
+#[derive(Debug, Serialize, Deserialize, ToSchema, Clone, Eq, PartialEq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+#[serde(rename_all = "snake_case")]
+pub enum ServiceProbeSuccess {
+    /// Successful result from a connectivity probe.
+    Connected,
+    /// Successful result from a probe specific to a Kafka service.
+    Kafka(KafkaServiceProbeSuccess),
+}
+
+/// Range of possible errors that can occur during a service probe.
+/// These are shared across all services.
+#[derive(ThisError, Debug, Serialize, Deserialize, ToSchema, Clone, Eq, PartialEq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub enum ServiceProbeError {
+    #[error("Timeout was exceeded")]
+    TimeoutExceeded,
+    #[error("Request type ({request_type:?}) was issued to a service type ({service_type}) which does not support it")]
```

Comment:
Always prefer active voice: `Service type ({service_type}) does not support request type ({request_type:?})`